lib/compat/wordpress-6.8/site-editor.php

```diff
@@ -0,0 +1,153 @@
+		return add_query_arg( array( 'p' => '/wp_block/' . $_REQUEST['postId'] ), remove_query_arg( array( 'postType', 'postId' ) ) );
+	}
+
+	if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
+		return add_query_arg( array( 'p' => '/pattern' ) );
+	}
+
+	if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
+		return add_query_arg( array( 'p' => '/wp_template_part/' . $_REQUEST['postId'] ), remove_query_arg( array( 'postType', 'postId' ) ) );
+	}
+
+	// The following redirects are for backward compatibility with the old site editor URLs.
+	if ( isset( $_REQUEST['path'] ) && '/wp_template_part/all' === $_REQUEST['path'] ) {
+		return add_query_arg(
+			array(
+				'p'        => '/pattern',
+				'postType' => 'wp_template_part',
+			),
+			remove_query_arg( 'path' )
+		);
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/page' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/page' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/wp_template' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/template' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/patterns' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/pattern' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/navigation' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/navigation' ), remove_query_arg( 'path' ) );
+	}
+
+	return add_query_arg( array( 'p' => '/' ) );
+}
+
+function gutenberg_redirect_site_editor_deprecated_urls() {
+	$redirection = gutenberg_get_site_editor_redirection();
+	if ( false !== $redirection ) {
+		wp_redirect( $redirection, 301 );
+		exit;
+	}
+}
+add_action( 'admin_init', 'gutenberg_redirect_site_editor_deprecated_urls' );
+
+function gutenberg_redirect_posts_dataviews_to_post() {
+	global $pagenow;
+	if (
+		'admin.php' !== $pagenow ||
+		! isset( $_REQUEST['page'] ) ||
+		'gutenberg-posts-dashboard' !== $_REQUEST['page'] ||
+		isset( $_REQUEST['p'] )
+	) {
+		return;
+	}
+
+	wp_redirect( gutenberg_get_posts_dataviews_url( '/' ), 301 );
+	exit;
+}
+
+add_action( 'admin_init', 'gutenberg_redirect_posts_dataviews_to_post' );
+
+
+/**
+ * Filter the `wp_die_handler` to allow access to the Site Editor's new pages page
+ * for Classic themes.
+ *
+ * @param callable $default_handler The default handler.
+ * @return callable The default handler or a custom handler.
+ */
+function gutenberg_styles_wp_die_handler( $default_handler ) {
+	if ( ! wp_is_block_theme() && str_contains( $_SERVER['REQUEST_URI'], 'site-editor.php' ) && isset( $_GET['p'] ) ) {
+		return '__return_false';
+	}
+	return $default_handler;
+}
+add_filter( 'wp_die_handler', 'gutenberg_styles_wp_die_handler' );
```

Comment:
I think this fix deserves some detailed explanation in a comment. It looks completely magical to me, I don't know what it does, I don't even know what a "hybrid theme" is.

Reply:
haha, you're right. Basically, site-editor.php's access is forbidden for hybrid/classic themes and only allowed with some very special query args (some very special pages like template parts...). This is done in Core and the only way to disable it if we change urls in Gutenberg is to override the `wp_die_handler`. (We did this in the past when we introduced the patterns page in classic themes too).

I think in the backport PR of this PR, I'm going to try to remove that blocking condition entirely from Core cause I believe it doesn't really make much sense anymore. 

But I'll add a comment.

Reply:
Is the last comment good enough or you think I should modify it? I'm doing the backporting btw since now it's mandatory to open a PR before being able to merge this PR into Gutenberg :)

Reply:
Now when I know what's going on I'm personally satisfied 🙂 And it's written down in GitHub comments, a motivated researcher can track it down.